### ISSUE TITLE
Proposal: add event hooks

### ISSUE THREAD

**Issue by cmccoy** [GoogleCloudPlatform/PerfKitBenchmarker#205]
## TL;DR
Some bugs (e.g. #146), and some additional desired functionality (e.g., annotating samples with host load information) could be fairly easily addressed by adding event hooks to PKB.

I suggest the following:
* Add a new module: `perfkitbenchmarker.events`.
* Define a class `EventDispatcher` within that module with methods:
   * `Subscribe(event_name, function)`
     Adds function to subscriber list for `event_name`.
   * `Unsubscribe(event_name, function)`
     Removes `function` from subscriber list to `event_name`.
   * `FireEvent(event_name, *args, **kwargs)`
     Logs event, calls all listeners of `event_name` with `*args`, `**kwargs`.
* Create a singleton `EventDispatcher` for the application as a whole.
  Listeners will be registered there, and events fired.

## Issues to think about:

* Threading.  Multiple benchmarks may be running simultaneously; multiple threads may be running for a single benchmark.
* How will we document the interfaces for event handlers (expected arguments)?
* Can events have side effects?
   * Internally: definitely yes (e.g., populating metadata for a metadata provider).
   * Otherwise: likely no.

## Useful events

This is a (non-comprehensive) list of events what could enhance the product.

#### Arguments to listener:

* Benchmark
* Flag name
* Flag value

#### Triggered when:

A benchmark uses a command-line flag.

#### Listeners:

* `FlagMetadataProvider`: Adds distinct used flags and their values to sample metadata.

#### Arguments to listener:

* Stage (benchmark, prepare, run, cleanup)
* benchmark_spec.

#### Triggered when

A run phase (prepare, run, cleanup) starts/ends.

#### Listeners:

   * A function/class that creates Sample objects with end-to-end time for each phase?

### `MachineReachable`

*Low priority.*

#### Arguments to listener:

* Virtual machine

#### Triggered when

The first `hostname` command succeeds on a VM.

#### Listeners:

* Cluster boot benchmark - gives time to boot before apt-get update runs.

Comments welcome!

**Comment by ivansmf** [comment]
Can we design this to also handle streamed results to BigQuery? How is an event different from a sample besides the listener end point?

**Comment by voellm** [comment]
Ivan ... Perf your request you could register and even handler to send events to BQ.  

The intent here however is not on performance timing but as an extension point.

**Comment by voellm** [comment]
Sorry I was a bit slow on the review but  I think there is one missing feature to make this 10x more useful and that is a flag to dynamically load even handlers from the command lline like this...

./pkb.py --evenhandler=xperf-trace.py --benchmarks=...

This would allow me to extend the functionality of PKB which having to change around the codebase.

Just a thought...

**Comment by cmccoy** [comment]
Closed by ec38eadb50a5b5009ef5955de77baf09d999e962. @voellm - can we open a separate issue for loading handlers from a file? Plugins will add a fair bit of complexity.
